corgidrp/l4_to_tda.py

```diff
@@ -137,6 +137,267 @@ def determine_color_cor(input_dataset, ref_star, source_star):
+            # Repeat single value for all frames
+            slit_per_frame = slit_per_frame * len(spec_dataset)
+    else:
+        slit_per_frame = [slit_transmission] * len(spec_dataset)
+
+    history_messages = []
+
+    for idx, frame in enumerate(spec_dataset):
+        if 'SPEC' not in frame.hdu_list:
+            raise ValueError("Input dataset does not contain a 'SPEC' extension.")
+
+        spec = frame.hdu_list['SPEC'].data.astype(float, copy=True)
+        spec_header = frame.hdu_list['SPEC'].header
+        spec_err = frame.hdu_list['SPEC_ERR'].data.astype(float, copy=True)
+
+        if spec_header.get('BUNIT', '').strip().lower() != "photoelectron/s":
+            raise ValueError("SPEC extension must have BUNIT 'photoelectron/s' before flux calibration.")
+
+        # Apply slit transmission correction if requested
+        slit_vals = slit_per_frame[idx]
+        slit_applied = False
+        if slit_vals is not None:
+            slit_vals = np.asarray(slit_vals, dtype=float)
+            if slit_vals.shape != spec.shape:
+                raise ValueError(f"slit_transmission array shape {slit_vals.shape} must match SPEC data shape {spec.shape}.")
+            
+            # Divide by slit transmission, handling division by zero
+            nonzero_mask = slit_vals != 0
+            spec = np.divide(spec, slit_vals, out=np.full_like(spec, np.nan), where=nonzero_mask)
+            spec_err = np.divide(spec_err, slit_vals, out=np.full_like(spec_err, np.nan), where=nonzero_mask)
+            
+            slit_applied = True
+            spec_header['SLITCOR'] = True
+        else:
+            spec_header['SLITCOR'] = False
+
+        # Apply flux calibration factor and color correction
+        color_cor_fac = frame.ext_hdr.get('COL_COR', 1.0)
+        factor = fluxcal_factor.fluxcal_fac / color_cor_fac
+        factor_error = fluxcal_factor.fluxcal_err / color_cor_fac
+
+        # Convert to flux units and propagate uncertainties
+        spec_flux = spec * factor
+        spec_flux_err = np.sqrt((spec_err * factor) ** 2 + (spec * factor_error) ** 2)
+
+        frame.hdu_list['SPEC'].data[:] = spec_flux
+        frame.hdu_list['SPEC_ERR'].data[:] = spec_flux_err
+        spec_header['BUNIT'] = "erg/(s*cm^2*AA)"
+        if 'SPEC_ERR' in frame.hdu_list:
+            frame.hdu_list['SPEC_ERR'].header['BUNIT'] = "erg/(s*cm^2*AA)"
+
+        history_messages.append(
+            f"Calibrated 1D spectrum with fluxcal_factor={fluxcal_factor.fluxcal_fac}, "
+            f"COL_COR={color_cor_fac}, slit_correction={slit_applied}."
+        )
+
+    if history_messages:
+        spec_dataset.update_after_processing_step(
+            " ".join(history_messages),
+            header_entries={"SPECUNIT": "erg/(s*cm^2*AA)", "FLUXFAC": fluxcal_factor.fluxcal_fac}
+        )
+
+    return spec_dataset
+
+
+def compute_spec_flux_ratio(host_dataset, companion_dataset, fluxcal_factor, 
```

Comment:
I'm not sure there is always a single corresponding host star file to each companion file. It may be cleaner to just pass in a single L4 host spectrum image and a single L4 companion spectrum image.